### PR TITLE
api: search: deprecate is_automated field, and is-automated filter

### PR DESCRIPTION
**Pull request by thaJeztah**
- related: https://github.com/docker/cli/pull/4413

### api: search: deprecate is_automated field, and is-automated filter

The is-automated field is being deprecated by Docker Hub's search API, and will always be set to "false" in future.

This patch deprecates the field and related filter for the Engine's API. In future, the `is-automated` filter will not yield any results when searching for `is-automated=true`, and will be ignored when searching for `is-automated=false`.

Given that this field is deprecated by an external API, the deprecation will not be versioned, and will apply to any API version.

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

```
api: search: deprecate "is_automated" field, and "is-automated" filter
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: api/types/registry/registry.go
Comment (corhere) [inline api/types/registry/registry.go:98]: Making the field `omitempty` is a breaking change for some clients which don't use Go's JSON parser. We should be nice to third-party clients by only omitting the field in responses to requests with the latest API version, leaving `"is_automated": false` in responses for older API versions.

**Review by corhere — CHANGES_REQUESTED** [review]
"An external service will no longer be setting this field to true" is not a valid excuse to make unversioned schema-breaking changes to API responses.

File: api/types/registry/registry.go
Comment (thaJeztah) [reply api/types/registry/registry.go:98]: Ah, you're right yes.

Comment (thaJeztah) [reply api/types/registry/registry.go]: updated; PTAL

File: api/server/router/image/image_routes.go
Comment (corhere) [reply api/server/router/image/image_routes.go]: Couldn't we version this change by e.g. making it an invalid request to search using an `is-automated` filter condition on API v1.45+?

> so we should just ignore the filter, and reset the field

We can ignore `is-automated=false` as that will match all images, but we can't just ignore `is-automated=true`. It would be more surprising for a client searching with an `is-automated=true` filter to get results where `"is_automated": false` than to get no results at all.

File: docs/api/version-history.md
Comment (corhere) [inline docs/api/version-history.md:28]:
```suggestion
```

**Comment by thaJeztah** [comment]
@corhere updated; removed the part about "no longer be returned" in future; PTAL

File: docs/api/version-history.md
Comment (corhere) [inline docs/api/version-history.md:31]: https://github.com/moby/moby/pull/45925#discussion_r1259980160

**Review by corhere — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
@corhere I updated the descriptions with (I think) we came to during the maintainers call

- In future, the field will always be "false"
- Because of that, searching for `is-automated=true` will not yield any results.
- And searching for `is-automated=false` will have no effect (so effectively "ignored")

Let me know if that looks good to you.

File: docs/api/version-history.md
Comment (corhere) [inline docs/api/version-history.md:32]: I still find the phrasing awkward. It reads like Docker Hub is directly deprecating a part of the Engine API, which doesn't make sense. They are distinct APIs with independent specifications; their only relationship is that one happens to be _implemented_ in terms of the other.

> Deprecated: The `is_automated` field in the `GET /images/search` response has been deprecated and will always be set to false in the future _because Docker Hub is deprecating the `is_automated` field in its search API._ The deprecation _is_ not versioned, and _applies_ to all API versions.
> Deprecated: The `is-automated` filter for the `GET /images/search` endpoint. The `is_automated` field has been deprecated by Docker Hub's search API. Consequently, searching for `is-automated=true` will yield no results. The deprecation is not versioned, and applies to all API versions.

**Review by corhere — APPROVED** [review]
I don't see any mention of the deprecation in https://docs.docker.com/docker-hub/release-notes/

I'm still not entirely happy with the version history entries, but my remaining nitpicks aren't serious enough to continue blocking this PR.

**Review by vvoland — APPROVED** [review]

**Review by laurazard — APPROVED** [review]
LGTM

File: docs/api/version-history.md
Comment (thaJeztah) [reply docs/api/version-history.md]: Thanks! I updated with your suggestion 👍

File: docs/api/version-history.md
Comment (corhere) [inline docs/api/version-history.md:26]: A rogue underscore snuck in. Hope it doesn't mess up the markdown rendering too much

Comment (thaJeztah) [reply docs/api/version-history.md:26]: Ah, derp! ❤️ 😂 

Looks like GitHub is fine showing it as-is `s_`, but now I can't un-see; I'll fix it in post 😂 

Screenshot 2023-08-02 at 09 08 42

Comment (thaJeztah) [reply docs/api/version-history.md:26]: - Here goes: https://github.com/moby/moby/pull/46143

**Pull request merged**
